File: 1029-VerticalOrderTraversalOfABinaryTree/1029-VerticalOrderTraversalOfABinaryTree.cpp
// Last updated: 11/2/2025, 10:29:20 PM
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> verticalTraversal(TreeNode* root) {
        map<int,map<int,multiset<int>>>nodes;
        queue<pair<TreeNode*,pair<int,int>>>todo;
        todo.push({root,{0,0}});
        while(!todo.empty())
        {
            auto curr = todo.front();
            todo.pop();
            TreeNode* node = curr.first;
            int x = curr.second.first;
            int y = curr.second.second;
            nodes[x][y].insert(node->val);
            if(node->left!=NULL)
            {
                todo.push({node->left,{x-1,y+1}});
            }
            if(node->right!=NULL)
            {
                todo.push({node->right,{x+1,y+1}});
            }
        }
        vector<vector<int>>ans;
        for(auto n:nodes)
        {
            vector<int>col;
            for(auto subn:n.second)
            {
                col.insert(col.end(),subn.second.begin(),subn.second.end());
            }
            ans.push_back(col);
        }
        return ans;
    }
};

File: 1029-VerticalOrderTraversalOfABinaryTree/1029-VerticalOrderTraversalOfABinaryTree.java
// Last updated: 11/2/2025, 10:29:16 PM
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
import java.util.*;
class Solution {

    class Info {
        TreeNode root;
        int col;
        int row;
        public Info(TreeNode root, int col, int row) {
            this.root = root;
            this.col = col;
            this.row = row;
        }
    }
    public List<List<Integer>> verticalTraversal(TreeNode root) {
        TreeMap<Integer, List<int []>> map = new TreeMap<>();
        Queue<Info>q = new LinkedList<>();
        q.add(new Info(root,0,0));

        while(!q.isEmpty()){
            Info curr = q.remove();

            int col = curr.col;
            int row = curr.row;
            int value = curr.root.val;

            map.putIfAbsent(col, new ArrayList<>());
            map.get(col).add(new int[]{row, value});

            if(curr.root.left != null){
                q.add(new Info(curr.root.left, curr.col - 1, curr.row + 1));
            }

            if(curr.root.right != null){
                q.add(new Info(curr.root.right, curr.col + 1, curr.row + 1));
            }

        }

        List<List<Integer>> ans = new ArrayList<>();
        for( List<int[]> colValues : map.values()){
                        Collections.sort(colValues, (a, b) -> a[0] == b[0] ? Integer.compare(a[1], b[1]) : Integer.compare(a[0], b[0]));

            List<Integer> sortedCol = new ArrayList<>();
            for(int [] pair : colValues){
                sortedCol.add(pair[1]);
            }
            ans.add(sortedCol);
        }
        return ans;
    }
}